Re-host Nexus on vm

jjgriff93 opened the pull request:
# PR for issue

## What is being addressed

In order to enable #1089 and proxy Docker Hub, we need to be able to map multiple ports to the Nexus docker container so that Docker hub can have its dedicated port. This means that we need to move away from App Service which limits us to only being able to map one custom port and instead host Nexus on a VM to give us the flexibility we require.

## How is this addressed

- Removed App Service TF
- Added Nexus VM definition 
- Initialise Nexus container with cloud-init startup script
- Rewrote nexus properties config and admin password reset scripts
- Added private DNS zone and vnet linking to allow connectivity to Nexus Vm from core and workspace vnets using DNS name
- Integrated lets encrypt SSL cert as part of cloudinit bootstrapping, placing it in nexus keystone and configuring Nexus Jetty instance to consume cert and handle traffic via SSL

Merged in #1480 to enable SSL cert integration so that user resources can access secure proxy without any additional configuration.

Fixes #1609

### Update
Need to add 
- [x] caching to the PR/Main builds to prevent lets encrypt blacklisting (repeated challenge attempts instead of reading letsencrypt local metadata 
- [x] CRON job in Nexus VM to renew certificate on daily basis
- [ ] 

Closes #1479 and closes #1481

jjgriff93 commented:
/test

jjgriff93 commented:
/test

jjgriff93 commented:
/test

marrobi reviewed (commented):
Deployed this, fixed a couple of things, but can connect to Nexus.

Couple of comments:
1. The scripts in ```cloud-config.yaml``` might be better as script files that are pulled into cloud config?
2. Can the nexus bootstrap steps be included in the cloud-config.yaml, as an alternate to running them manually?

@tamirkamara are you able to take a look over this code wise as fresh pair of eyes? Thanks.

Great work.

tamirkamara reviewed (commented):
Great progress on this @jjgriff93.
My main concern is that the api won't be able to deploy nexus anymore which is the way forward for all shared services. I believe this will also block @tanya-borisova from moving the nexus deployment to an extended e2e test.
I thought of merging nexus-cert into the main nexus dir/bundle and treat them both as a single component. Was that something you considered? Will that play nicely with issuing the certificate from within the bundle?

jjgriff93 commented on .github/workflows/deploy_tre_reusable.yml:
```diff
@@ -452,6 +452,13 @@ jobs:
+      - name: Cache Nexus Letsencrypt metadata
+        uses: actions/cache@v3
+        with:
+          key: nexus letsencrypt
```
Nexus now not deployed as part of cicd so removed cache step

jjgriff93 commented:
> Deployed this, fixed a couple of things, but can connect to Nexus.
> 
> Couple of comments:
> 
> 1. The scripts in `cloud-config.yaml` might be better as script files that are pulled into cloud config?
> 2. Can the nexus bootstrap steps be included in the cloud-config.yaml, as an alternate to running them manually?
> 
> @tamirkamara are you, or one of your team able to take a look over this code wise as fresh pair of eyes? Thanks.
> 
> Great work.

Thought the same re having the cloud-init bash scripts as separate bash files and invoke via the yml however I couldn't find any clean way of copying scripts from the terraform agent into the vm to then invoke via cloud-init. If you know of a way we can do this let me know and happy to implement.

As for the Nexus bootstrap steps, completely agree and will add :)

jjgriff93 commented on Makefile:
```diff
@@ -147,8 +152,30 @@ prepare-tf-state:
 	&& pushd ./templates/core/terraform > /dev/null && ../../shared_services/firewall/terraform/remove_state.sh && popd > /dev/null \
 	&& pushd ./templates/shared_services/firewall/terraform > /dev/null && ./import_state.sh && popd > /dev/null
+
+terraform-shared-service-deploy:
+	$(call target_title, "Deploying ${DIR} with Terraform") \
+	&& . ./devops/scripts/check_dependencies.sh \
+	&& . ./devops/scripts/load_env.sh ./templates/core/.env \
+	&& . ./devops/scripts/load_env.sh ./devops/.env \
+	&& . ./devops/scripts/load_terraform_env.sh ./devops/.env \
+	&& . ./devops/scripts/load_terraform_env.sh ./templates/core/.env \
+	&& . ./devops/scripts/key_vault_list.sh \
+  && if [[ "$${TF_LOG}" == "DEBUG" ]]; then echo "TF DEBUG set - output supressed - see tflogs container for log file" && cd ${DIR} && ../../deploy_from_local.sh 1>/dev/null 2>/dev/null; else cd ${DIR} && ../../deploy_from_local.sh; fi;
```
Good spot, seems to have been left over from Olly's merge. Removed.

jjgriff93 commented on templates/shared_services/sonatype-nexus/terraform/cloud-config.yaml:
```diff
@@ -0,0 +1,186 @@
+    groups: [docker]
+
+write_files:
+  - path: /etc/cron.hourly/docker-prune
+    # An hourly cron job to have docker free disk space. Running this frquently
+    # since disk might get full fast, but we prune only when free space is low.
+    content: |
+      #!/bin/bash
+      set -o errexit
+      used_percent=$(df / --output=pcent | tail -1 | sed 's/[^0-9]//g')
+      echo "Used disk space percent: $${used_percent}"
+      if (( used_percent > 60 )); then
+        echo "Free space too low, pruning..."
+        docker system prune -f
+      fi
+    permissions: '0755'
```
Yep you're right, not really needed here as its a single docker image. Removed.

jjgriff93 commented on templates/shared_services/sonatype-nexus/terraform/cloud-config.yaml:
```diff
@@ -0,0 +1,186 @@
+  # Set up Nexus to serve https using SSL cert
+  - path: /home/adminuser/configure-nexus-ssl.sh
+    content: |
+      # Prepare ssl certificate
```
Added

jjgriff93 commented on templates/shared_services/nexus-cert/Dockerfile.tmpl:
```diff
@@ -0,0 +1,29 @@
+    && curl -sL https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor | tee /etc/apt/trusted.gpg.d/microsoft.gpg > /dev/null \
+    && AZ_REPO=$(lsb_release -cs) \
+    && echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ $AZ_REPO main" | tee /etc/apt/sources.list.d/azure-cli.list \
+    && apt-get update && apt-get -y install azure-cli
```
Added

jjgriff93 commented on templates/shared_services/nexus-cert/Dockerfile.tmpl:
```diff
@@ -0,0 +1,29 @@
+
+ARG BUNDLE_DIR
+
+RUN apt-get update && apt-get install -y ca-certificates
```
Added

jjgriff93 replied:
Added

jjgriff93 commented:
> Great progress on this @jjgriff93.
> My main concern is that the api won't be able to deploy nexus anymore which is the way forward for all shared services. I believe this will also block @tanya-borisova from moving the nexus deployment to an extended e2e test.
> I thought of merging nexus-cert into the main nexus dir/bundle and treat them both as a single component. Was that something you considered? Will that play nicely with issuing the certificate from within the bundle?

Yes a good point - originally this PR was written before the shared services moved to the new deployment model. The reason we have the cert service separate from Nexus is because down the line we want to re-use the cert service to generate certs for other services, not just Nexus. However I think we can at least merge in the letsencrypt invocation so it's ran as part of the bundle - then expect the cert service bundle to be registered and deployed beforehand as a dependency

jjgriff93 commented on templates/shared_services/nexus-cert/terraform/appgateway.tf:
```diff
@@ -0,0 +1,163 @@
+  lifecycle { ignore_changes = [tags] }
+}
+
+resource "azurerm_application_gateway" "agw" {
```
I think a better approach for this would be as part of executing the logic currently in make nexus-letsencrypt, this is done as a custom action on the bundle, which takes care of starting and stopping the service while it generates the cert. That way its only up and running while generating a cert and put into stand-by afterwards

tamirkamara replied:
ok, so I assume this is also something you're working on...

jjgriff93 commented on Makefile:
```diff
@@ -93,7 +93,7 @@ firewall-install:
 	&& $(MAKE) shared-service-register-and-deploy DIR=./templates/shared_services/firewall/ BUNDLE_TYPE=shared_service
 
-nexus-install:
+nexus-install: nexus-cert-install nexus-letsencrypt
```
Currently looking at using a custom porter action to do this logic within the bundle so it can be handled via the API

tamirkamara replied:
Please note that thr RP (porter) doesn't have access to the internet currently...

jjgriff93 replied:
I think that without this we can't deploy this bundle via the API in a self-contained way, the letsencrypt portion will have to be ran manually beforehand via the deploying agent that has internet connectivity. Is there a story for adding internet connectivity to the RP? If so I can mention that and create a new issue where we can then implement lets encrypt as part of the bundle

tamirkamara replied:
@jjgriff93, How about this idea?
Resource processor sits in its own subnet which is governed by a separate firewall rule. You could create a similar rule to [this one](https://github.com/microsoft/AzureTRE/blob/main/templates/shared_services/firewall/terraform/firewall.tf#L132) with the domain lets encrypt requires. One could also add this domain/s to the existing rule but I thought a new one tied to the deployment of Nexus is better (which can be done is a separate task)

BTW, in the future, when multistep deployments will be available we might have a different way to accomplish it.

jjgriff93 replied:
That's worked :) have implemented a custom action on the certs shared service called 'generate' which can be used to generate and renew a certificate from the API

jjgriff93 commented on templates/shared_services/sonatype-nexus/terraform/vm.tf:
```diff
@@ -0,0 +1,156 @@
+  scope                = data.azurerm_key_vault.kv.id
+  role_definition_name = "Key Vault Reader"
+  principal_id         = azurerm_user_assigned_identity.nexus_msi.principal_id
+}
```
It was needed to give the msi a scope within the subscription otherwise the az cli login would fail, but now realised we can get around the with the `--allow-no-subscriptions` flag so have removed

jjgriff93 commented:
Added new story for improvements to the re-usability of the certs shared service: #1756

jjgriff93 commented:
@marrobi I've managed to figure out a way to add the configuration scripts to the VM and invoke them after Nexus is bootstrapped, so we now have a self-contained shared service that configures itself with no need for the jump box :) also have separated out the cloud-init scripts into bash files as requested

jjgriff93 commented:
@tamirkamara the certs service is now fully self-contained and invocable via the API - it accepts two parameters of the fqdn to generate the cert for and the name to save the cert as in KeyVault. It runs the letsencrypt script within the resource processor as part of a "generate" custom action, allowing it be renewed as and when needed. I've suggested further improvements to the certs service for re-usability in https://github.com/microsoft/AzureTRE/issues/1756 which is out of scope for this PR

jjgriff93 commented:
/test

ross-p-smith commented on docs/tre-admins/setup-instructions/configuring-shared-services.md:
```diff
@@ -2,21 +2,81 @@
+## Configure Nexus
 
-## Configure Nexus repository
+Before deploying the Nexus shared service, you need to make sure that it will have access to a certificate to configure serving proxies over https with. By default, the Nexus service will serve proxies from `https://nexus-{TRE_ID}.{LOCATION}.cloudapp.azure.com/`, and thus it requires a certificate that validates ownership of this domain to use for SSL.
```
```suggestion
Before deploying the Nexus shared service, you need to make sure that it will have access to a certificate to configure serving secure proxies. By default, the Nexus service will serve proxies from `https://nexus-{TRE_ID}.{LOCATION}.cloudapp.azure.com/`, and thus it requires a certificate that validates ownership of this domain to use for SSL.
```

ross-p-smith commented on docs/tre-developers/letsencrypt.md:
```diff
@@ -0,0 +1,52 @@
+
+The additional resources are as followed:
+
+1. Public IP provisioned in the same location as the web app the certificate is intended for with a domain lable which matches the web apps name.
```
```suggestion
1. Public IP provisioned in the same location as the web app that the certificate is intended for; this will also have a domain label which matches the web app name.
```

ross-p-smith commented on docs/tre-developers/letsencrypt.md:
```diff
@@ -0,0 +1,52 @@
+
+1. Public IP provisioned in the same location as the web app that the certificate is intended for; this will also have a domain label which matches the web app name.
+1. Storage Account with a static web app.
+1. Application gateway to route traffic from thepPublic IP to the static web app
```
```suggestion
1. Application gateway to route traffic from the Public IP to the static web app
```

ross-p-smith reviewed (approved):
I've locked at the code and it seems to be mainly around the Shared Services, so I am going to assume that if Marcus is happy - I am!

marrobi commented on docs/tre-admins/setup-instructions/configuring-shared-services.md:
````diff
@@ -2,21 +2,81 @@
+1. Run the below commands in your terminal to build, publish and register the Nexus shared service bundle:
+
+  ```cmd
+  make bundle-build DIR=./templates/shared_services/certs
````
@ross-p-smith should these be the nexus bundle not certs?

marrobi commented on docs/tre-admins/setup-instructions/configuring-shared-services.md:
````diff
@@ -2,21 +2,81 @@
+1. Run the below commands in your terminal to build, publish and register the Nexus shared service bundle:
+
+  ```cmd
+  make bundle-build DIR=./templates/shared_services/certs
+  make bundle-publish DIR=./templates/shared_services/certs
+  make bundle-register DIR=./templates/shared_services/certs BUNDLE_TYPE=shared_service
````
```suggestion
  make bundle-build DIR=./templates/shared_services/sonatype-nexus
  make bundle-publish DIR=./templates/shared_services/sonatype-nexus
  make bundle-register DIR=./templates/shared_services/sonatype-nexus BUNDLE_TYPE=shared_service
```

marrobi commented:
I had an issue with this as have firewall rules with the priorities used in the file. This is where @damoodamoo's work is needed. I have adjusted my existing rules for now.

However now I'm getting keyvault issues:
```
ng to request: StatusCode=409 -- Original Error: autorest/azure: Service returned an error. Status=409 Code=\"Conflict\" Message=\"Certificate nexus-ssl is currently in a deleted but recoverable state, and its name cannot be reused; in this state, the certificate can only be recovered or purged.\" InnerError={\"code\":\"ObjectIsDeletedButRecoverable\"}\u001b[0m \u001b[31m│\u001b[0m \u001b[0m \u001b[31m│\u001b[0m \u001b[0m\u001b[0m  with azurerm_key_vault_certificate.tlscert, \u001b[31m│\u001b[0m \u001b[0m  on certificate.tf line 10
```

Does the certificate name need to be unique? Have the shared service ID in it?

marrobi reviewed (changes requested):
Two issues:
- Clashing firewall priorities - I suggest we move the priorities up into the high hundreds until the firewall updating PR is in.
- Issues with nexus SSL certs and keyvault - the cert name needs to be unique for each install - might just be one to add to the docs for now?

jjgriff93 commented:
> Two issues:
> 
> * Clashing firewall priorities - I suggest we move the priorities up into the high hundreds until the firewall updating PR is in.
> * Issues with nexus SSL certs and keyvault - the cert name needs to be unique for each install - might just be one to add to the docs for now?

Have addressed: firewall rule no. updated to 601 and added note to docs re kv conflicts and how to manage

jjgriff93 commented:
/test

jjgriff93 commented:
/test

marrobi commented on docs/tre-admins/setup-instructions/configuring-shared-services.md:
```diff
@@ -1,22 +1,112 @@
+
+Just bear in mind that if this service is redeployed any changes in the UI won't be persisted. If you wish to add new repositories or alter existing ones, use the JSON files within the `./nexus_repos_config` directory.
+
+### B. Migrate from existing Nexus service (hosted on App Service)
```
Don't think this will work as the workspace will be missing the DNS zone.

jjgriff93 replied:
Have added a note requiring workspace version `0.2.14` or above

marrobi reviewed (changes requested):
Couple of docs things, I will test deployment of nexus alongside a new workspace with new VMs.

jjgriff93 commented on docs/tre-admins/setup-instructions/configuring-shared-services.md:
```diff
@@ -1,22 +1,112 @@
+## Deploy/configure Nexus
 
-Make sure you run the following command using git bash and set your current directory as C:/AzureTRE
+There is a new Nexus shared service which can be located in the `./templates/shared_services/sonatype-nexus-vm` directory, with the bundle name `tre-shared-service-sonatype-nexus`, which is now hosted using a VM to enable additional configuration required for proxying certain repositories.
```
Moved to its own section below set up guide

jjgriff93 commented:
/test

jjgriff93 commented:
/test

marrobi commented on templates/core/terraform/resource_processor/vmss_porter/main.tf:
```diff
@@ -176,5 +176,6 @@ resource "azurerm_key_vault_access_policy" "resource_processor" {
 
-  secret_permissions = ["Get", "List", "Set", "Delete", "Purge", "Recover"]
+  secret_permissions      = ["Get", "List", "Set", "Delete", "Purge", "Recover"]
+  certificate_permissions = ["Get", "Recover"]
```
Think this needs import too.

jjgriff93 replied:
Yep you're right, added

marrobi commented on templates/shared_services/certs/terraform/appgateway.tf:
```diff
@@ -0,0 +1,168 @@
+  location            = data.azurerm_resource_group.rg.location
+
+  sku {
+    name     = "Standard_v2"
```
If stopping and starting nto sure its an issue.

jjgriff93 commented:
/test

marrobi reviewed (approved):
Tested  other than the VM flag and all good.

jjgriff93 commented:
/test
